src/backend/distributed/executor/adaptive_executor.c

```diff
@@ -1577,22 +1578,135 @@ AcquireExecutorShardLocksForExecution(DistributedExecution *execution)
 		return;
 	}
 
+	bool parallelExecutionNotPossible =
+		list_length(taskList) == 1 || ShouldRunTasksSequentially(taskList);
+
+	bool anyAnchorTableIsReplicated = AnyAnchorTableIsReplicated(taskList);
+	if (!anyAnchorTableIsReplicated && parallelExecutionNotPossible)
+	{
+		/*
+		 * When a distributed query on tables with replication
+		 * replication factor == 1, we rely on Postgres to handle the
+		 * serialization of the concurrent operations on the workers.
+		 *
+		 * For reference tables, even if their placements are replicated
+		 * ones (e.g., single node), we acquire the distributed execution
+		 * locks to be consistent when new node(s) are added.
+		 */
+		return;
+	}
+
 	/*
-	 * When executing in sequential mode or only executing a single task, we
-	 * do not need multi-shard locks.
+	 * We first assume that all the remaining modifications are going to
+	 * be serialized. So, start with an ExclusiveLock and lower the lock level
+	 * as much as possible.
 	 */
-	if (list_length(taskList) == 1 || ShouldRunTasksSequentially(taskList))
+	int lockMode = ExclusiveLock;
+
+	if (!anyAnchorTableIsReplicated && IsCoordinator())
 	{
-		Task *task = NULL;
-		foreach_ptr(task, taskList)
+		/*
+		 * When all writes are commutative then we only need to prevent multi-shard
+		 * commands from running concurrently with each other and with commands
+		 * that are explicitly non-commutative. When there is no replication then
+		 * we only need to prevent concurrent multi-shard commands.
+		 *
+		 * In either case, ShareUpdateExclusive has the desired effect, since
+		 * it conflicts with itself and ExclusiveLock (taken by non-commutative
+		 * writes).
+		 *
+		 * However, some users find this too restrictive, so we allow them to
+		 * reduce to a RowExclusiveLock when citus.enable_deadlock_prevention
+		 * is enabled, which lets multi-shard modifications run in parallel as
+		 * long as they all disable the GUC.
+		 *
+		 * We also skip taking a heavy-weight lock when running a multi-shard
+		 * commands from workers, since we currently do not prevent concurrency
+		 * across workers anyway.
+		 */
+		lockMode =
+			EnableDeadlockPrevention ? ShareUpdateExclusiveLock : RowExclusiveLock;
+	}
+
+	if (AllModificationsCommutative ||
+		(parallelExecutionNotPossible && modLevel < ROW_MODIFY_NONCOMMUTATIVE))
```

Comment:
is there a name we can give to `modLevel < ROW_MODIFY_NONCOMMUTATIVE` ? I'm not sure what's lower.

(previously this was addressed by super elaborate comments & and 1 condition at a time)

Reply:
it might be slightly nicer to separate this into if and else and elaborate on each condition